Framework: Move localStorage polyfill to separate bundle

aduth opened the pull request:
As discovered in #7038 and #6558 before it, our local storage polyfill loading behavior is very fragile, and I suspect we're doomed to repeat it in its current state. As such, this pull request seeks to separate the polyfill into a separate bundle to be included in the page prior to the main application bundle.

**Testing instructions:**

Verify that your primary browser and Safari Private Mode load Calypso without any errors in the console.

**Open questions:**
- Are there desktop app considerations to be made here?
- Why does the polyfill bundle include vendor code? See related webpack/webpack#1016 and [examples/two-explicit-vendor-chunks](https://github.com/webpack/webpack/tree/master/examples/two-explicit-vendor-chunks) for reference
- **very important:** do we need systems to whitelist `public/polyfill.js` since it’s a static asset?

/cc @gwwar (fixed previous breakage) @blowery (tinkerer of Webpack)

Test live: https://calypso.live/?branch=add/polyfill-bundle

gwwar commented on client/lib/local-storage/test/index.js:
```diff
@@ -5,43 +5,41 @@ import { assert } from 'chai';
 describe( 'localStorage', function() {
 	describe( 'when window.localStorage does not exist', function() {
-		const window = {};
+		delete global.localStorage;
```
It feels like we should stub, or restore this value when the suite is done.

gwwar commented:
This line is causing the failed tests: https://github.com/Automattic/wp-calypso/blob/master/client/lib/react-test-env-setup/index.js#L31

gwwar commented:
cc @mkaz for the desktop question

aduth commented:
>This line is causing the failed tests:

Indeed, though interestingly different failures when updating this line. We might instead need to change `lib/local-storage` back into exporting a function, though this would require additional logic in the server rendered markup to invoke. I'll plan to poke at this more tomorrow.

aduth commented:
Rebased and introduced a separate `lib/polyfill` module to better allow `lib/local-storage` to be used outside the context of the browser (e.g. test env setup) and more easily tested in isolation. Having a separate `lib/polyfill` module better aligns with "polyfill" as a bundle, rather than specifically to the localStorage polyfill (in case, for example, we want to include more polyfills, which hopefully is not the case 😄 ).

gwwar commented:
Whoops lost track of this one. I can test again after a rebase

aduth commented:
Thanks @gwwar , pushed a rebase to resolve merge conflicts.

I noticed that the conflicts were caused by the introduction of the [ES6 function wrappers](https://github.com/Automattic/wp-calypso/blob/003eb1132b467f704e30a35c1570765910ebc1ac/client/boot/index.js#L1-L5) in #6117, which feels kinda similar in purpose. Do you think it would make sense to pull that into the "polyfill" bundle here? To be fair, those are more development mode utilities than polyfills. I dunno, maybe best to leave for a separate pull request in any case.

gwwar commented:
>  Do you think it would make sense to pull that into the "polyfill" bundle here?

Let's leave that for another PR

gwwar commented:
With a completely empty `lib/local-storage/index.js` file the polyfill chunk is still `385 kB` in size in a development environment. Do we expect this amount of size overhead from webpack in the first chunk? cc @blowery

gwwar commented:
👍 I confirmed that browsing with private Safari browser does not lead to errors in Calypso on this branch. 

The polyfill chunk size seems unreasonably large though, so we may want to investigate if an alternative webpack configuration can give us a lighter result.

blowery commented:
@gwwar hmmm.. if it's an entry chunk, it'll include the loader bits from webpack, but 385 seems... crazy. What all is in there?

aduth commented:
When building as production, it appears to be much smaller - Around 30kb without any minification or gzip (~5.5kb after). Inspecting generated bundle, most of this size is CoreJS polyfills for ES6 e.g. `Object.assign`.

gwwar commented:
30kb sounds much more reasonable. I'd be comfortable with merging this.

gwwar commented:
Checking in on this one, was there anything else that needed additional testing?

aduth commented:
>Checking in on this one, was there anything else that needed additional testing?

No, just haven't had a chance to circle back around to this. Will need to rebase and retest to account for changes in #8161.

aduth commented:
Rebased to account for #8161. @blowery : Do you think this should be included in the vendor bundle? Currently it's separate.

Also, I'm not sure why it was needed previously, but I did not include the `CommonsChunkPlugin` updates as in the previous version (9036fffea97c637ecb2cc3525e427d05dd0ce1a3) and it still appears to work well. In fact, the development bundle size is much more reasonable, so that could explain why it was so large before.

blowery commented on client/lib/polyfill/index.js:
```diff
@@ -0,0 +1,12 @@
+import localStoragePolyfill from 'lib/local-storage';
+
+export default ( () => {
+	if ( 'undefined' === typeof window ) {
```
Hrm, just realized I may not understand how modules evaluate as well as I thought. Does code in the body of the module run before the module is `require`d? Or just once on the first `require`?

aduth replied:
I think you're right that we don't need or want the `export default`. The only other reason to keep it an IIFE is to allow the early return, since there can't be top-level returns in a module. Otherwise we could just put all the browser-specific polyfills inside the `if` block.

>Or just once on the first `require`?

This is the intended behavior.

aduth replied:
Removed IIFE export in ec9331998b78e1236af9bb7f2b20478f83c33d2b.

blowery commented:
re 
> Also, I'm not sure why it was needed previously, but I did not include the CommonsChunkPlugin updates as in the previous version (9036fff) and it still appears to work well. In fact, the development bundle size is much more reasonable, so that could explain why it was so large before.

Could you walk through that a bit more? I'm curious why the dev bundle would change in size at all.

aduth commented:
@blowery To be honest, I don't recall why I felt the need to include it in `CommonsChunkPlugin` in the first place. Probably because I was referencing the approach for vendor. When rebasing, I "forgot" to include it again, yet it ran without issues and the size of the bundle was appropriate in development, so... ¯\\_(ツ)_/¯

lancewillett commented:
@aduth Want to keep this one active?

We plan close it soon as it is an older pull with no recent activity.

aduth commented:
@lancewillett Yes, I'd still like to merge this in. It's reviewable in its current state as far as I can recall.

gwwar commented:
I'd be fine with 🚢'ing this after a rebase and another spot check.

mtias commented:
@aduth is this good to go?

aduth commented:
@blowery Do you see any reason this might conflict with your ongoing work in #11867 ?

blowery commented on server/pages/index.jade:
```diff
@@ -118,13 +118,15 @@ html(lang=lang, dir=isRTL ? 'rtl' : 'ltr', class=isFluidWidth ? 'is-fluid-width'
 		if i18nLocaleScript
 			script(src=i18nLocaleScript)
 		if 'development' === env || isDebug
+			script(src=urls[ 'polyfill' ])
```
does this work? I would suspect it has to come _after_ manifest? Though I guess no one requires it, so ...

blowery commented on webpack.config.js:
```diff
@@ -95,6 +95,8 @@ if ( CALYPSO_ENV === 'desktop' || CALYPSO_ENV === 'desktop-mac-app-store' ) {
 	// no chunks or dll here, just one big file for the desktop app
 	webpackConfig.output.filename = '[name].js';
 } else {
+	webpackConfig.entry.polyfill = [ path.join( __dirname, 'client', 'lib', 'polyfill' ) ];
```
i'd probably bump this out to a separate webpack build entirely. This will be affected by things like the vendor dll / vendor chunk, commons plugins, etc. I know it has no deps right now, but if could be a footgun in the future.

aduth replied:
Eugh, makes me wonder if we should just write it in plain JS and source it as a static resource from `public` or WPCOM.

blowery replied:
that would be fine. it's just a polyfill.

blowery commented:
@aduth I don't see an issue with that, but I'm curious why this works at all. I would expect the polyfill to be a separate webpack build entirely, so it's excluded from the commons chunk plugins and the dll.

blowery commented:
This could use a rebase too. It's based on a pretty old commit and has to do some auto-merge magic to apply.

blowery commented:
> but I'm curious why this works at all

<img width="734" alt="following_ _reader_ _wordpress_com" src="https://cloud.githubusercontent.com/assets/14350/23757354/947aa69e-04b4-11e7-9695-a3d82873301b.png">

It doesn't work as it. The polyfill needs `__webpack_jsonp` defined and that happens in manifest.

Also, digging through the polyfill built code, it's requiring `debug`, which is part of the manifest chunk that's output. 

So moving it after manifest may be enough to get it going.

aduth commented:
>It doesn't work as it.

Is this after a rebase? May very well be that this pull request hadn't been updated since #8161 .

blowery commented:
Hrm, nope, the polyfill depends on something in manifest that depends on vendor. So the polyfill has to come after vendor to work. Hrm. Is that too late?

blowery commented:
> Is this after a rebase?

Yup

> May very well be that this pull request hadn't been updated since #8161 .

Yup, likely.

aduth commented:
Rebased and tried static file idea in a02ff84 . If we were to do this, it might make more sense to have the script live in the WordPress.com codebase to take advantage of automatic minification.

blowery commented on public/polyfill.js:
```diff
@@ -0,0 +1,32 @@
+	var _data = {};
+
+	// We can't replace window.localStorage, but we can replace its methods
+	window.localStorage.setItem = function( id, val ) {
```
since we're not minifying this, you could stash `window.localStorage` into a `var s` and assign against that. Save a few bytes. :) 

Though maybe gzip eats it too.

blowery commented on server/pages/index.jade:
```diff
@@ -115,6 +115,7 @@ html(lang=lang, dir=isRTL ? 'rtl' : 'ltr', class=isFluidWidth ? 'is-fluid-width'
 				}
 			})();
 
+		script(src='/calypso/polyfill.js')
```
should append the hash here, just in case we have to change it. maybe tie into the other static bits in `index.js`?

aduth replied:
>should append the hash here, just in case we have to change it. maybe tie into the other static bits in `index.js`?

We could add it at the point a change needs to be made, which speaks to a bigger problem of: if anyone were to need to make a change, would they realize they'd also need to update the cache flag? ...things we take for granted with the build process 😄

gziolo replied:
You can calculate the hash based on size of the file or its last modification date. We can skip that, too. I don't think this is going to change in the future.

blowery commented:
> If we were to do this, it might make more sense to have the script live in the WordPress.com codebase to take advantage of automatic minification.

That'll make local dev without a network pretty rough.

aduth commented:
Meh, static file has a number of downsides. Not sure that outweighs the need to create a separate config for such a small file though?

blowery commented:
Yeah. I'd consider just inlining it honestly.

aduth commented:
That's a good thought. In which case it acts more as an extension to what we have already for BrowseHappy?

blowery commented:
That sounds cool.

aduth commented:
My reluctance at this point stems from wariness to add more unminified JavaScript to the base `index.jade` file. The BrowseHappy stuff is already pretty non-ideal.

>i'd probably bump this out to a separate webpack build entirely. This will be affected by things like the vendor dll / vendor chunk, commons plugins, etc. 

Can you explain what you mean by this being affected @blowery ?

blowery commented:
Just that it'll have to load after the manifest if its part of our normal webpack build. That's probably ok?

gziolo commented:
My question is why we don't migrate existing code to use localForage instead of localStorage?

aduth commented:
>My question is why we don't migrate existing code to use localForage instead of localStorage?

We should, but it's non-trivial since the leap from synchronous to asynchronous is not always well-defined. Since it's not obvious, also noting that most usage won't be uncovered from a search of `localStorage` but instead the `store` dependency we use to interface.

gziolo commented:
I did some research and it looks like `localforage` supports callbacks and ES6 Promises:
* https://github.com/mozilla/around/blob/90805796a0612460b18d2b5b99fd4d625bb83f08/www/js/vendor/backbone.localforage.js#L28
* https://github.com/mozilla/around/blob/90805796a0612460b18d2b5b99fd4d625bb83f08/www/js/views/users.coffee#L32

They mention callback and promises also [here](https://hacks.mozilla.org/2014/02/localforage-offline-storage-improved/).

gziolo commented:
I did even more research and it looks like we would need sync wrapper for `localforage` that would work as drop-in replacement for `store` :D

I think your PR makes much more sense in this context.

blowery commented:
yeah, there's a few places using it:
```
$ ag "require\( 'store'" -c && ag "from 'store'" -c
boot/index.js:1
layout/community-translator/launcher.jsx:1
layout/nux-welcome/index.js:1
lib/application-passwords-data/index.js:1
lib/desktop/index.js:1
lib/local-list/index.js:1
lib/plugins/store.js:1
lib/plugins/wporg-data/list-store.js:1
lib/posts/actions.js:1
lib/preferences/actions.js:1
lib/products-list/index.js:1
lib/signup/progress-store.js:1
lib/signup/flow-controller.js:1
lib/trophies-data/index.js:1
lib/user/user.js:1
my-sites/welcome/README.md:1
auth/controller.js:1
blocks/app-promo/index.jsx:1
layout/community-translator/invitation-utils.js:1
layout/masterbar/notifications.jsx:1
lib/abtest/index.js:1
lib/countries-list/index.js:1
lib/oauth-token/index.js:1
lib/olark/index.js:1
lib/sites-list/list.js:1
lib/user/support-user-interop.js:1
lib/wp/handlers/guest-sandbox-ticket.js:1
my-sites/invites/controller.js:1
my-sites/invites/invite-accept-logged-out/index.jsx:1
my-sites/pages/helpers.js:1
reader/sidebar/index.jsx:1
```

gziolo commented:
@aduth I rebased it with master. 

@fabianapsimoes Can you test this branch with the scenario described in https://github.com/Automattic/wp-calypso/issues/6555#issuecomment-299468069?

klimeryk commented:
With https://github.com/Automattic/wp-calypso/pull/13748 merged, this can probably be closed - unless we want to pursue this as the ultimate fix? Although, IMHO, https://github.com/Automattic/wp-calypso/pull/13748 seems solid enough (though, I'm probably biased ;)).

aduth commented:
@klimeryk Sounds alright to me! We can always revisit if needed.
